Serialize Distance and Percentiles aggregate using an invariant locale

rickardp opened the pull request:
See issue #4041

Added unit tests to show problem and added a fix using the invariant culture to serialize the type. I only saw the problem with fractional values but there can potentially also be problems with thousand separators in some locales.

russcam commented on src/Nest/CommonOptions/Geo/Distance.cs:
```diff
@@ -7,7 +7,7 @@
 {
 	[JsonFormatter(typeof(DistanceFormatter))]
-	public class Distance
+	public class Distance : IFormattable
```
`IFormattable` shouldn't be implemented; the string representation of `Distance` should always use `CultureInfo.InvariantCulture`

rickardp replied:
Makes sense if string representation is only used for serialization, which I assume it is.

russcam commented on src/Tests/Tests/CommonOptions/DateMath/DateMathExpressions.doc.cs:
```diff
@@ -232,5 +232,21 @@ public class DateMathExpressions
 			(oneYear > fiftyTwoWeeks).Should().BeTrue();
 		}
+		[U]
+		public void DifferentLocales()
```
Can use `UseCultureAttribute` on test method, and separate out the three different culture tests into separate unit tests. Since these are in a `*.doc.cs` file, these should also be marked with `// hide` to not be emitted in generated documentation e.g.

```c#
// hide
[U]
[UseCulture("sv-SE")]
public void SerializesWithInvariantCulture()
{
   // ..implementation
}
```

rickardp replied:
Cool stuff! Didn't know about that!

rickardp replied:
Now I see, the UseCultureAttribute is copied from xunit samples and available only in Tests.Reproduce. I will move the unit test there unless you think it's a bad idea @russcam

russcam replied:
ah, gotcha. I think moving it to `Tests.Core` project, perhaps under `Xunit` directory might make most sense? That way, both Tests.Reproduce and Tests can use it

russcam commented on src/Tests/Tests/CommonOptions/DistanceUnit/DistanceUnits.doc.cs:
```diff
@@ -101,5 +102,36 @@ public void UsingDifferentUnits()
 			Expect("45.5nmi").WhenSerializing(new Distance(45.5, Nest.DistanceUnit.NauticalMiles));
 		}
+		[U]
+		public void DifferentLocales()
```
Can use `UseCultureAttribute` on test method, and separate out into separate unit tests. Since these are in a `*.doc.cs` file, these should also be marked with `// hide` to not be emitted in generated documentation e.g.

```c#
// hide
[U]
[UseCulture("sv-SE")]
public void SerializesWithInvariantCulture()
{
   // ..implementation
}
```

rickardp replied:
I tried to follow the pattern of the other GH issue repro cases where the issue number is part of the unit test file name.

russcam commented on src/Tests/Tests/LocaleUtil.cs:
```diff
@@ -0,0 +1,30 @@
+
+namespace Tests
+{
+	internal sealed class LocaleUtil : IDisposable
```
Can remove this and use `UseCultureAttribute`

rickardp commented:
When I moved the unit test I found another unit test that was failing if the culture is not en-US (GithubIssue3715.CanDeserializeExtendedStats). Will investigate

russcam commented:
Feel free to open a separate issue for GithubIssue3715.CanDeserializeExtendedStats. I can take a look at it (tomorrow) :+1:

rickardp commented:
Created issue #4044 regarding the issue with CanDeserializeExtendedStats. The problem is probably the `double.Parse` calls in `AggregateFormatter` (from just briefly looking at the stacktrace of the failed test).

Might be a good idea to scan the code for `<numeric type>.Parse` without a CultureInfo.InvariantCulture in the serialization code.

rickardp commented:
I added the fix for issue #4044 as well since I think it was a no-brainer. In AggregateFormatter, I added InvariantCulture to the double.Parse call. I did a full solution search for double.Parse and the only remaining case where InvariantCulture is not used is in StringToDoubleConverter in Utf8Json, and it looks like a fall-back case.

russcam commented:
ported to 7.x
